core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -979,6 +1004,21 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
         }
     }
 
+    private void verifyPrimary() {
+        if (shardRouting.primary() == false) {
+            // must use exception that is not ignore be replication logic. See TransportActions.isShardNotAvailableException
+            throw new UnsupportedOperationException("shard is not a primary " + shardRouting);
+        }
+    }
+
+    private void verifyReplicationTarget() {
+        final IndexShardState state = state();
+        if (shardRouting.primary() && shardRouting.isRelocationTarget() == false && state != IndexShardState.RELOCATED) {
```

Comment:
This bikeshed is going to be very multi-colored by the time we are done with it. :smile: 

+1 for @ywelsch's proposal, but I do not like `UnsupportedOperationException` here but `IllegalStateException` would be fine. I think of `UnsupportedOperationException` as being reserved for cases where the receiving object does not support the invoked method, but `IllegalStateException` as the receiving object rejects the invoked method because of the current state.

:angel: